dots and modal features

Dusan-Mojsilovic-EY opened the pull request:
Added dots and modal features with some style.

Dusan-Mojsilovic-EY committed:
dots and modal features

milica-roljic-ey commented on src/Components/Carousel/Carousel.js:
```diff
@@ -2,14 +2,27 @@ import React, {useEffect, useState} from 'react';
 import './Carousel.scss';
 
```
Remove empty line here. Let's keep import statements grouped at the top of the file without adding new line between them.

milica-roljic-ey commented on src/Components/Carousel/Carousel.js:
```diff
@@ -2,14 +2,27 @@ import React, {useEffect, useState} from 'react';
 import './Carousel.scss';
 
 import useFetch from '../useFetch/useFetch';
+import Modal from '../Modal/Modal'
```
We can end statements with or without `;` . Good practice is to be consistent. Decide how you want to write statements and stick to it.

milica-roljic-ey commented on src/Components/Carousel/Carousel.js:
```diff
@@ -25,7 +38,19 @@ const Carousel = () => {
+        }
+    })
+
+    console.log(text)
```
Remove `console.log()`

Dusan-Mojsilovic-EY replied:
Meni ni jedan ovaj console.log name u radi na racunaru, nisam siguran dal sam ja poslao pravi ili sta se desilo.

milica-roljic-ey commented on src/Components/Carousel/Carousel.js:
```diff
@@ -35,22 +60,30 @@ const Carousel = () => {
+        if ( offset < numberReq - count) {
         setOffset(offset + count)
-    }
+        }
```
Code formatting should be improved.
Check out this article https://medium.com/@pgivens/write-cleaner-code-using-prettier-and-eslint-in-vscode-d04f63805dcd

milica-roljic-ey commented on src/Components/Carousel/Carousel.js:
```diff
@@ -61,17 +94,25 @@ const Carousel = () => {
+        {activePage === 1 ? <button className="active"></button> : <button className="btnStyle"></button>}
+        {activePage === 2 ? <button className="active"></button> : <button className="btnStyle"></button>}
+        </div>
+        <Modal modal={modal}
```
Let's write props in new lines and closing tag to be in line with opening tag.
            <Modal
                modal={modal}
                data={text}
                handleClose={closeModal}
            />

milica-roljic-ey commented on src/Components/Carousel/Carousel.js:
```diff
@@ -35,22 +60,30 @@ const Carousel = () => {
-
     return(
-        <div className="Carousel">
+        <div className='Container'>
```
className should always have value written in kebab case style, not title case. This should be className='container' and etc. Please update all classess in js and scss files.

milica-roljic-ey commented on src/Components/useFetch/useFetch.js:
```diff
@@ -5,8 +5,8 @@ import { useState, useEffect} from 'react';
     var apiKey = "jbfndddBALH0QmDvRJIzbDOChFEdT9PoSH08CeUV";
-    var count = 20;
-    var url =`https://api.nasa.gov/planetary/apod?api_key=${apiKey}&count=${count}`
+    var numberReq = 20;
```
It's not an error to use var to declare and initialize variables, but I see that you are already using let and const. I would use const for these two variables numberReq and url.

milica-roljic-ey replied:
Also apiKey. Missed that in previous comment.

milica-roljic-ey commented on src/Pages/Home/Home.js:
```diff
@@ -10,7 +10,7 @@ const Home = () => {
           <div className="ApplicationPart">
               <p>Lorem ipsum dolor sit amet consectetur adipisicing elit. Maxime cumque distinctio ducimus, repellendus similique eos sed corporis obcaecati. Soluta nesciunt fugiat quo expedita alias, in quas distinctio neque quam id?</p>
-              <button className="AppButton">Start Application Process</button>
+              <button id="AppButton">Start Application Process</button>
```
Same comment here, just for ids, instead classNames.

milica-roljic-ey commented on src/Components/Footer/Footer.scss:
```diff
@@ -1,18 +1,21 @@
 .Footer {
-    position: fixed;
+    // position: fixed;
```
Is this commented out accidentally? 1. Footer is not sticky without this line of code 2. If we think that some property is not needed, let's remove it, instead of commenting.

milica-roljic-ey commented on src/Components/useFetch/useFetch.js:
```diff
@@ -16,7 +16,7 @@ const useFetch = () => {
             console.log(res.data)})
```
Remove `console.log()`

Milica-Roljic-EY commented on src/Components/Modal/Modal.scss:
```diff
@@ -0,0 +1,58 @@
+    font-size: 16px;
+    border-radius: 15px;
+    font-family: "Ubuntu";
+        .Modal-Information{
```
Check indentation.

milica-roljic-ey commented on src/Components/Carousel/Carousel.js:
```diff
@@ -25,7 +38,19 @@ const Carousel = () => {
+        } else if (offset + count == numberReq) {
+            setActivePage(2)
+        }
+    })
```
useEffect contains a call to 'setActivePage'. Without a list of dependencies, this can lead to an infinite chain of updates. To fix this, pass [offset, numberReq, count] as a second argument to the useEffect Hook.

milica-roljic-ey commented on src/Components/Carousel/Carousel.js:
```diff
@@ -61,17 +94,25 @@ const Carousel = () => {
                     <img src={e.hdurl} className="ImageWrapper"/>
```
img elements must have an alt prop, either with meaningful text, or an empty string for decorative images. You can use e.title or even empty string is fine.

milica-roljic-ey commented:
Hi Dusan, good work so far! Please see my comments in PR and below:

Pay more attention to code formatting, clean and well structured code is easier to read and debug.
I'm confused with file Responsive.js. Custom hook from that file is not used anywhere, return statement has typo and it's not exported properly. Please take a look and update it.
Some components have imports that are not used (e.g. Footer)
Constants (like api keys, or strings that are visible to the users) we can keep in separate files. (Optional for now)
Let's think about how we can make this carousel more customized.

I think this PR should be updated before merging to master.

Dusan-Mojsilovic-EY committed:
comments-updated

milica-roljic-ey commented on package.json:
```diff
@@ -12,6 +12,7 @@
     "react-dom": "^17.0.2",
     "react-router-dom": "^5.2.0",
     "react-scripts": "4.0.3",
+    "uuid": "^8.3.2",
```
This package is not needed in order to add unique key for list. This is used e.g. for cookies. Please remove it from package.json.

milica-roljic-ey commented on src/Constants.js:
```diff
@@ -0,0 +1,3 @@
+export const apiKey = "jbfndddBALH0QmDvRJIzbDOChFEdT9PoSH08CeUV";
+export const numberReq = 20;
+export let url =`https://api.nasa.gov/planetary/apod?api_key=${apiKey}&count=${numberReq}`
```
Why is this export let?

milica-roljic-ey commented on src/Components/useFetch/useFetch.js:
```diff
@@ -16,7 +12,7 @@ const useFetch = () => {
             console.log(res.data)})
```
Remove console.log

milica-roljic-ey commented on src/Components/Carousel/Carousel.js:
```diff
@@ -1,78 +1,133 @@
+import "./Carousel.scss";
+import useFetch from "../useFetch/useFetch";
+import Modal from "../Modal/Modal";
+import { v4 as uuidv4 } from 'uuid'
```
See comment above for uuid.

milica-roljic-ey commented on src/Components/Carousel/Carousel.js:
```diff
@@ -1,78 +1,133 @@
-            <div className="RightButton">
-                <button id="forward" onClick={goForward}>⇨</button>
-            </div>
+        <div className="curouselPictures">
```
Check spelling here for className

milica-roljic-ey commented on src/Components/Carousel/Carousel.js:
```diff
@@ -1,78 +1,133 @@
+  const [modal, setModal] = useState(false);
+  const [text, setText] = useState(null);
+
+  const showModal = (tt) => {
```
For parameter names, it is not obligatory, but it's very useful if it's more descriptive.

milica-roljic-ey commented on src/Components/Carousel/Carousel.js:
```diff
@@ -1,78 +1,133 @@
+  };
+
+  if (!data) {
+    return null;
```
Also you can return empty fragment <></> instead of null.

milica-roljic-ey commented:
Hi Dusan,
I have reviewed updated PR, and here's my comments:
1. Code formatting is fixed, it's more readable now, that's great. :) 
2. For lists and keys, it's fine if you use index as unique identifier - I would concatenate it with string that describes the list. 
3. Dots (bullets) are not working as expected. If you cannot implement it to show more than 3 when needed and to match position of the image, it's better to remove it.
4.  Footer is still not sticky
![Screenshot 2021-06-28 002041](https://user-images.githubusercontent.com/85943191/123561445-a35a1700-d7a8-11eb-89cb-0b86369f7a37.png)
5.  Take a look how you can style React components, and how it's correctly written https://egghead.io/lessons/react-style-react-components-with-classname-and-inline-styles, https://www.code-boost.com/beginners-guide-to-scss/ 
6.  Think about User experience when we are loading data for carousel - e.g. spinner (it's optional)

milica-roljic-ey reviewed (changes requested):
Hi Dusan, please see my comment in Conversation tab and update PR accordingly. Thanks!

Dusan-Mojsilovic-EY committed:
2nd review - changes updated

Dusan-Mojsilovic-EY commented:
Hi Milice,
I updated my project with comments you have sent me.

1. I removed uuid for keys and made my new keys with descriptions that come with all the pictures. I see no errors in console.
2. I removed dots as you requested. Will definitely try to add this feature in the future, but for now I did not want that to slow down my project.
3.  I took back sticky footer. This was primary implemented, but I had removed it. Now footer is sticky again.
4. I will definitely pay attention more on how to style React components in further work.
5.  I made spinner for user experience. This made me a lot of trouble, but at the end its working properly :)

I pushed all my changes on GitHub and have not made any new pull request, if I have understood correctly. Please let me know if I need to do anything else.

Thank you very much.

Dusan

mitrovic-jovana commented on src/Components/Carousel/Carousel.js:
```diff
@@ -36,15 +35,15 @@ const Carousel = () => {
+  //   } else if (offset + count == numberReq) {
+  //     setActivePage(2);
+  //   }
+  // }, [offset, numberReq, count]);
```
If commented code is not needed, it's better to remove it

Dusan-Mojsilovic-EY replied:
done

mitrovic-jovana commented on src/Components/Carousel/Carousel.js:
```diff
@@ -69,11 +68,13 @@ const Carousel = () => {
+  let newData;
 
-  const newData = data.slice(offset, offset + count);
+  if (data.length === 0) {
```
Good job on adding some indicator that data is still not loaded

mitrovic-jovana commented on src/Components/Carousel/Carousel.js:
```diff
@@ -84,12 +85,12 @@ const Carousel = () => {
             newData.map((e) => {
               return (
-                <div className="wrapper" key={uuidv4()}>
+                <div className="wrapper" key={e.explanation}>
```
This would be a good place to do some destructuring. Also, try to define variables with some meaningful name, "e" does not tell us what that variable represents. In most of the cases e is shorten of event, so it's a little bit confusing

mitrovic-jovana commented on src/Components/Modal/Modal.js:
```diff
@@ -9,7 +9,6 @@ const Modal = ({ modal, data, handleClose }) => {
     <div className="modal-Container">
       <section className="modal-Main">
```
Destructure data here

Dusan-Mojsilovic-EY replied:
done

Dusan-Mojsilovic-EY committed:
3rd review

Dusan-Mojsilovic-EY commented:
all changes updated and pushed

mitrovic-jovana reviewed (approved):
Hi Dusan, good job on following Milica’s suggestions and adding fixes. Carousel works as expected, I agree that you can focus on adding pagination if you have some spare time. I will merge your changes, however there is a space for improving the design and code structure. Here are a few suggestions that you can add at some point of time:
1.	I see you have the same button on few places. When you see repetitive code, that is a sign that you can move that to separate component. So, you can add Button as a separate component which could have props like event handler and background color. 
2.	Cards in carousel can also be separate component and you could use it as a child component in Carousel. That way you could reuse it in other places in the application
3.	Don’t invoke NASA’s service inside Carousel component. If you do that, Carousel is not customizable, and it can only work with response from that api. Invoke service in parent component and pass data to Carousel as props
4.	Since you don’t have pagination, one suggestion for better UX is to disable left arrow when you are at the beginning of the carousel and right arrow when you get to the end of carousel. Also, add pointer courser to the arrows
5.	I left you a comment for defining variables with some meaningful name, in Carousel component on line 15 you have variable “tt”. That really is not a good name. Go trough your code and check if there are other variables that could be renamed
6.	Think about folder structure organization. For example, I would move Images folder outside Components folder. Name of the folder should tell you what is inside it and images are not components. Move Constants.js to separate folder instead of putting it directly to src folder. Make folder for custom hook etc.
7.	Put links for Privacy Policy and Terms and Conditions next to each other in the footer
